db/version_builder.cc

```diff
@@ -514,10 +515,26 @@ class VersionBuilder::Rep {
     return Status::OK();
   }
 
-  void MaybeAddFile(VersionStorageInfo* vstorage, int level, FileMetaData* f) {
+  void MaybeInheritFile(VersionStorageInfo* vstorage, int level,
+                        FileMetaData* f) {
     if (levels_[level].deleted_files.count(f->fd.GetNumber()) > 0) {
       // f is to-be-deleted table file
       vstorage->RemoveCurrentStats(f);
+      deleted_base_files_[f->fd.GetNumber()] = f;
+    } else {
+      vstorage->AddFile(level, f, info_log_);
+    }
+  }
+
+  void MaybeAddFile(VersionStorageInfo* vstorage, int level, FileMetaData* f,
+                    uint64_t base_max_file_number) {
+    if (f->fd.GetNumber() <= base_max_file_number) {
+      auto it = deleted_base_files_.find(f->fd.GetNumber());
+      assert(it != deleted_base_files_.end());
```

Comment:
@yiwu-arbug @Connor1996 Do you think we need to set a background error when this assertion fails? Violation of this is almost equivalent to data loss.

Reply:
`assert` is disabled for non-debug, so better set a background error if possible

Reply:
Just notice this method is allowed to fail. Change to return status instead.